### ISSUE TITLE
Please add a CLI argument that shows me what the changes are

### ISSUE THREAD

**Issue by StoneCypher** [prettier/prettier#6069]
There are too many tutorials advising me that the way to handle prettier is to commit my code to git, then run `prettier --write && git diff`

That's not practical in a continuous integration scenario.  Parsing that is extremely complex, especially when git versions are taken into account.  If your linting is parallel, which is common in configurability scenarios, you've just opened up the ur-write-lock problem.

Most linting and style tools' default behavior is to give a list of the things they think are incorrect.  I am unable to figure out how to get `prettier` to do this.

Please add a mechanism to get `prettier` to explain itself.

# Desired behavior

```
prettier --list-problems

./dir/foo.source:
11:3: didn't do the thing
81:2: no tabs, you yahoo

./place/thing.apl:
51:5: this should be six lines
```

**Comment by lydell** [comment]
Hi!

Could you elaborate on _why_ you want this?

We already have `prettier --check` for CI. It tells you _which files_ need formatting. In my opinion, that's all you need. Run `prettier --write` on them and you're good to go.

Note: Looking at your desired output, it sounds like you're looking for [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier).

**Comment by StoneCypher** [comment]
> We already have prettier --check for CI. It tells you which files need formatting. In my opinion, that's all you need.

Being told what the files are, but not what the problems are, is immensely frustrating.  

.

> Note: Looking at your desired output, it sounds like you're looking for eslint-plugin-prettier.

Several members of my team have advocated leaving `prettier` in favor of the eslint plugin, yes, because eslint will tell us what it thinks is wrong

If we do that, what's going to happen is people are going to say "but as long as we can pick and choose rules," etc

.

> Run prettier --write on them and you're good to go.

I will not allow a tool to make changes to source that I have not seen.  Prettier has already introduced several bugs.

If our code linting tools won't tell us what the problems are, we have to use other tools (like inappropriate git commits) to sort out what prettier is even trying to do

Please consider giving us the features that every other tool like this has for a reason, and that basically every other tool uses as their default interaction

**Comment by j-f1** [comment]
> Prettier has already introduced several bugs.

If you can make a reproducible example for these cases, we’d love to hear about and fix them. You can also run Prettier with the `--debug-check` argument, which will verify that the meaning of your code hasn’t been changed.

**Comment by j-f1** [comment]
The reason we don’t provide descriptions of what needs to be changed is that’s not how Prettier works. Instead of having a bunch of rules and making sure the code follows them, we basically throw away your original formatting and replace it with our own.

**Comment by StoneCypher** [comment]
That thing where they explain that the reason they won't give you what you need is they think you shouldn't want it

**Comment by StoneCypher** [comment]
Sometimes our needs aren't what you expect, and an explanation won't change our needs

**Comment by StoneCypher** [comment]
I frequently have to spend ten minutes trying to figure out what formatting change Prettier wants, because I can't just learn what's normal, because it won't tell me what it expects, meaning that at every single commit I have to stop and try to figure out what's going on because someone installed a commit hook that wastes my time.  Sometimes I lose my entire train of thought because there's an extra space in front of a comment somewhere.

If you can't tell me what the rules are, the rules shouldn't exist.

Because I can't learn your expectations, my flow is constantly broken.

I understand that you don't agree with the thing I need.  You should give it to me anyway.  It's almost no work for you, and this ruins my experience with your tool.

**Comment by StoneCypher** [comment]
look.

i'll be direct.  i don't actually want prettier.  it's something my teammates want, because they think that a formatting tool will help prevent arguments.

and that's fine.  i'm okay with that.

i tried really hard to get us on `eslint-plugin-prettier`, because it's acceptable to a normal workflow.  there's a problem?  fine, it'll tell you what it is, ***so that you won't do it next time***.

"but let it write for you."  

i don't want to.  i want to learn the problem so that i don't constantly have this minor interruption.  it's buggy and everyone knows it, which is why your own manual says "be careful with write," but then you advise write for basic usage in threads where people are asking you for quality of life features.

"but put it on a git commit hook."

***i don't want to.***

i want this tool to work like other tools.  i want it to tell me the rules.  i want to be able to learn the rules.  i don't want to constantly be interrupted for five minutes because there's a space in front of a comment and now i've lost my flow because i didn't even know what it wanted me to do.

how was i supposed to know that putting an extra line at the end of the file was a faulting defect?  it wouldn't just tell me that

formatting tools are fine, but this one is written by people who refuse to let you just learn what you're expected to do.

this is a quality of experience issue.  sometimes people want different things than you do.

i get that you're trying to normalize source code, but why are you trying to control my development flow?  that's going much too far.  my development flow doesn't affect the other developers.

this is converting from "the end result must look like this" to "you must work this way," and that's a non-productive change

i get that you don't want to do this thing, but it would be really easy for you, people keep asking for it, there are tutorials on how to work around this

can you please just accept that some people want to work a different way than you expect, and give them the things they need?

i shouldn't have to screw around get simple answers on what my expectations are

i don't want to get into a fight with my team to remove this

can you please just be flexible

**Comment by StoneCypher** [comment]
the only real problem i have with prettier is that i can't in a practical way learn the rules over time

when someone installs, say, an eslint rule set, then when there's a fault, it tells me what it is.  within a week or two, i've learned the local rules, and it's not an issue anymore.

i think formatting norms have value.  however, it's also worth recognizing that being faulted on any ruleset, be it a formatting set, a testset, et cetera, costs time, and breaks a person out of their workflow.  some, myself included, might even go as far as to suggest that that negative reinforcement cycle is a training method to get people to stop screwing around and follow norms.

the problem that i face is fourfold.

.

# i actually need this

it's not practical for me to walk you through what bugs i've faced.  we all know tools like this can fault.  it happens to all of them.  it's just a fact of life.  your own manual warns a person to be careful with `--write`, so to act like it's surprising that a person wants to be careful around that seems really strange to me.  on top of that, i use a lot of source transformation tools, some of them not public, and sometimes i use things that it's not clear whether you actually support.  in the past i've had problems around novel javascript being pushed through babel plugins, or my own custom babylon stuff.  

i don't even remember a lot of them, and there will be new ones in the future, and most of them will be my fault and it's really hard to figure out what i did wrong, because lord knows what's wrong

moreover, those bugs are going to change with time.  i'm going to use different tools with time.  i'm mostly exiting babel in favor of typescript now.  i don't want to come back and file and issue and wait on community help every time.  i need to be able to move forwards when something happens.  life involves time pressure, and being able to see the source transformations is a critical and basic immediate need tool when something is going wrong.  i can't even figure out whether prettier is involved in a problem, or whether the problem is in an adjacent tool, without this.

all that aside, it's beside the point.  it doesn't matter what's going wrong; this is a developer quality of life issue.  i'm telling you that i'm facing a significant pain point and i want help.  "you shouldn't want that" is not going to solve my pain.  

i understand that you want me to follow along, to let the tool do what it's going to do, to trust the machine.  i don't want that.  even if your record and history is perfect, other tools like these aren't, and it makes me nervous.  i just want to be thorough, and if i can't be thorough, it's not going to feel safe to me.

and i get that you're going to say "then prettier is not the tool for you."  i actually agree with that, but i work on a team, and some of them really want this.

i am, mostly, okay with prettier.  it seems to do a good job.

but sometimes it does something i don't know about.  and that seems unnecessary to me.  it would really improve my quality of experience with the tool if i could just be a nervous person and look at what it's going to do before it does it.  i wish you'd let me.

by example, the other day, a developer checked something in that violated a prettier rule.  it was just a single extra space in front of a line comment.  i didn't even know that was a thing that got validated.

when i merged their work, my own work faulted, and it was actually really frustrating to figure out why, because i was in the middle of a merge, so i couldn't use git to work around knowing what was wrong, and because of the nature of the code, i couldn't use an online comparator because the source shouldn't be seen by third parties (and besides those are not a good experience to find it.)  

the way i ended up dealing with it was to create a new directory, clone the source over there, install the eslint ruleset version, check it with that different tool, find out "oh there's a faulted space," fix it, and get back to what i was doing.

it cost a lot of time.

it would be simple to give me a diagnostic that i could just say "okay so what's wrong," and then prettier could go "it's this thing over here," and i could fix it in an editor and run it again, get a green light, and feel really good about, and in control of, the whole process

i recognize that it's not your preference; so, don't make it the default

but it would really help me out a lot

being told to not want it is actually really frustrating :frown: .  my needs haven't changed.

.

# i want to learn to avoid these 

my preference is to just learn the local rules and not hit these in the first place.  it's easier for me.  hitting these bumps and having to figure out what's wrong using git is pretty labor intensive, and i lose track of what was on my mind as a result.

the reason that's usually easy is when i get it wrong, in other tools, is they'll just give me a list of what's wrong.  as i understand it, there's an eslint ruleset that roughly implements prettier, and then would give me the list that i want.  however, one of my teammembers is strongly against that, on grounds that eslint would then be moderately configurable, and we would get away from the one true way, and into arguments.  (i think this is weird because we didn't argue before, but, it's important to him.)

so, i would like prettier to say "this file should have no newline at the end."  that's fine, i can totally do that.  but i want to learn that, rather than to do it my own way then have it enforced.  there's a non-zero time cost, because i don't want to skip the verification step.

and if i could just `prettier --list-changes` then look at them, then i'd be like "okay that's fine," do a `--write`, and then not make the mistake in the future

but because there is no `--list-changes`, and because i am unwilling to accept non-audited source transformations, i have to go through a slow expensive frustrating process.

my actual preference would be to learn the rules through short cycle negative reinforcement, the way i do with `eslint` and test cases.

`--list-changes` would really, really improve my experience with prettier.  a whole lot.

.

# i don't want source transformers associated with source control

life is complicated.  i don't trust tools in general.  this is not a slight against you; it's more just an observation that tools have bugs.

certain things are high stakes.  the database is one of those.  i don't like modified databases, plugins, layers, third party tunings.  i don't care if it's a "well trusted layer" or a "storage engine used by bigco" or a whatever.  the database is high stakes.  i want vanilla.  i want no wrappers.  i want minimal magic.

i feel similarly about source control.  the stakes are too high.  my git environment is virtually unconfigured.  i set the user and password, the editor, the default newline mode, and otherwise it's pure vanilla.  no wrappers, no bash preconfig, no nothing.  i want to be able to sit down at anyone's machine and just get the thing fixed.

to that end, i am allergic to git hooks of all forms.

i understand that they have value and that they improve a lot of developers' lives.  i just don't want them personally.  i'd much rather do that stuff in scripts and in continuous integration.

sometimes when you talk to a c programmer, one of the things they might tell you that they value is the absolute clarity of the language.  there's no overloading, no automatic stuff.  the entire behavior is what you're looking at.  there's no contextual knowledge, you don't have to look up any classes; it's just the code in front of you and the functions that calls.

and you know, i really like higher level languages, and when i do most of my work i work in them.  but if i had to do something life critical, i'd use something nearly context-less, in that sense.

so i accept that `--write` and hooks and automation are the right workflow for a whole lot of people

they aren't for me.  not even if someone tells me "do it this way."  i need to audit the changes that have been made in order to feel good about the results.

if i could just look at the changes first, then `--write` would become a wonderful lazy man's tool for getting the work done at low effort.  that's exactly the kind of thing that i want.

.

# prettier can change my source, but please let my workflow be my own

my current understanding of the value of prettier as primarily presented is that the normalization of source code formatting improves the experience for the team.  i can accept that.  my own private formatting is (ahem) exotic, and i've known some people whose formatting actually kind of bugged me, so, i can see the potential for my formatting to bug others.  that's fine.  let's get to a norm, and let's have a tool to do it the easy way.  sounds great.

but that's different than my workflow.

i am by nature at work very detail oriented.  i have an emotional need to be aware of what's happening in the source base.  it's actually just fine for my code to be changed, as long as i know what happened.

telling me to not want something isn't going to change who i am.

this will remain a pain point for me.  i really wish you would reconsider.  the effort is low, and the impact on people who themselves don't use it would probably be zero.

.

# it's okay to give people tools you don't personally want

it really just raises the quality of the tool, you know?

a lot of people really want it.  there's a bunch of tutorials online for how to work around this.  it doesn't cause an expense to workflows not using it.  there's no overhead or cost.

its value is highest when someone doesn't actually understand what's going on, and wants to exclude prettier as a cause, so even from the purist viewpoint of "we'd like to make prettier be a flawless transformer," even if it actually *is* a flawless transformer, if i can't see what it's doing, i can't leverage that knowledge to exclude it

one important form of pluralism is helping different kinds of people find value in the same tool

it is common with some nervous folk (by example myself) to just not trust automation in general.  for them, just being able to look at the changes on the way through before they're applied can actually be a pretty important thing.

i also read every single line of every single pr.  i also read every build log.  it's just my nature.

people are different and have different needs.  please be inclusive and flexible.

.

# understand that this is hard

it's really, really hard to ask more than once, especially after someone said no, especially when it involves talking about yourself.  i'm probably not going to be able to ask again after this.

i will always need this, even if someone tells me "you shouldn't want that" a whole lot of times.

please give me the help that i need

**Comment by j-f1** [comment]
Thank you for giving us more detail!

I think your teammate might be misunderstanding how Prettier integrates with ESLint. When you use `eslint-config-prettier`, all it does is disable all ESLint rules that check formatting. `eslint-plugin-prettier` runs Prettier directly, then does a diff between the input and the output and reports any differences (running ESLint with `--fix` will fix them automatically).

That should work if you’re only or primarily using Prettier for JavaScript/TypeScript. However, if you frequently use Prettier for other kinds of files, check out @ikatyang’s [`prettylint`](https://github.com/ikatyang/prettylint).

Another thing I’ll say is that many cases of Prettier apparently breaking code are actually Prettier showing that the code won’t do what it seems to do by pointing out syntax mistakes and correctly interpreting it.

Finally, once you get more used to using Prettier, I highly recommend installing an integration for whatever editor you use. This will allow Prettier to format your code when you save, which completely changes the coding experience. Instead of trying to conform to a style, you can just write your code out on one line, copy-paste without properly indenting, and more. Then, when you save, all the code will snap into the proper format. If it doesn’t, you’ll know there’s a syntax error somewhere even before you run a linter or tests.

**Comment by StoneCypher** [comment]
Please give me the tool that I need, instead of trying to explain me out of it 

No amount of explaining will change that I need this

**Comment by StoneCypher** [comment]
None of what you just said would have faced the example problem that I gave, or almost any of the others

I feel like I'm going to have to have a fight with my team and get this tool removed, because the tool authors won't accept that some people don't work the way they do, and create a tool that works for people who aren't like them

**Comment by StoneCypher** [comment]
None of what you just said would have faced the example problem that I gave, or almost any of the others

One of the worst experiences you can have as a developer is to be trapped between a team that ignores how much time a tool is wasting "because it will save time" and a tool author that ignores that you need a small feature that won't hurt anyone because "I feel like you don't understand your own needs"

I feel like I'm going to have to have a fight with my team and get this tool removed, because the tool authors won't accept that some people have different needs than they do

.

> I think your teammate might be misunderstanding how Prettier integrates with ESLint.

It doesn't matter.  I have to have a fight with them, at work, because you won't give me what I need.

**Comment by StoneCypher** [comment]
> Finally, once you get more used to using Prettier

My emotional needs go through my entire life, and will not change "once I get more used to using Prettier."

It is unfortunate that you are refusing to understand someone clearly saying "I am never, under any circumstances, comfortable with a tool like this if I can't see what it's doing."

**Comment by j-f1** [comment]
> None of what you just said would have faced the example problem that I gave, or almost any of the others

How does `prettylint` not solve your problem?

**Comment by StoneCypher** [comment]
i already explained that in great detail, and also explained why the emotional cost of repeating myself was too high

it's really weird that you've now proposed two alternate tools as surrogates for just making this one do what people actually need from it

i give up ☹️

**Comment by j-f1** [comment]
> i already explained that in great detail

I did a quick search of this page, and I don’t see you mentioning `prettylint`. Can you link me to that comment?

> it's really weird that you've now proposed two alternate tools as surrogates for just making this one do what people actually need from it

Prettier tries to follow the UNIX philosophy of doing one thing well. In this case, that’s taking some code and formatting it. That means that instead of adding lots of different features to the core Prettier project (thereby making it slower and larger for everyone), we instead make it easy for tools like `prettylint` to integrate with Prettier and add useful features.

**Comment by StoneCypher** [comment]
> Prettier tries to follow the UNIX philosophy of doing one thing well.

If people are coming to you begging you to solve their pain points, and what they're asking for is an audit of the thing it does, that means it's not succeeding.

**Comment by StoneCypher** [comment]
There are only two comments of substance here.

I'm exiting this thread now, because respectfully, it's pretty clear that you haven't actually read what I said yet.  You're talking about things entirely unrelated to my problem.

You could pretty much take your responses and put them on any request for help.  They don't interact with what I said at all.

**Comment by j-f1** [comment]
> because respectfully, it's pretty clear that you haven't actually read what I said yet

I did read everything you said. I’m sorry I wasn’t able to be more helpful.
